pythonwhat/State.py

```diff
@@ -164,13 +165,14 @@ def extract_function_calls(self):
 
         if (self.student_function_calls is None):
             fp = FunctionParser()
-            fp.imports = self.pre_exercise_imports
+            fp.imports = self.pre_exercise_imports.copy()
```

Comment:
Great find! FYI: this is not a deep copy... it's shallow copy (but that's all what's necessary here, since the values of the `self.pre_exercise_imports` are just strings). The original code was not even a copy, but rather a assigning the same reference to an object. Deep copy means you're recursively copying all objects in a dictionary (e.g. using `copy.deepcopy`, like it's used in `utils.copy_env`). This is not necessary here since `string` objects are immutable.